File: stackunderflow/etl/marts/project.py
"""project_mart — one row per project, lifetime totals.

Replace-from-scratch-for-affected-keys pattern. New events for an
existing project invalidate the prior aggregate (``total_sessions``
especially is a DISTINCT count that can't be summed across windows),
so we recompute the project row from all of its events. ``INSERT OR
REPLACE`` on the ``project_id`` PRIMARY KEY does the swap atomically.

``provider``, ``slug``, ``display_name`` come from the ``projects``
table, joined in by id.

Message-type + command dims
===========================
``usage_events`` is assistant-only (the normalizers skip non-billable
rows), so the token/cost/message totals above can't see user turns,
tool-result turns, or commands. The Overview's User / Assistant /
Tool-Use / Tool-Results cards and the Commands KPI need those, so we
materialise five extra columns (v022) computed straight off the
project's ``messages.raw_json`` — running the **same** classifier +
enricher field-extraction ``get_project_stats`` uses, so the counts are
identical to the full-pipeline aggregator (proven by the equivalence
tests). This is a build-time cost (a full ``messages`` scan per affected
project); the dashboard read path stays a single indexed ``project_mart``
lookup, well inside the <100ms budget.

The ``INSERT OR REPLACE`` above lists only the original 13 columns, so it
resets the dim columns to their ``DEFAULT`` on every refresh; the
``_refresh_message_dims`` second pass then UPDATEs the true counts for the
same affected projects (mirrors ``command_mart``'s session_count recompute).

Overview rate dims (v023)
=========================
The same second pass also materialises the numerators behind the Overview's
cache / interruption / errors blocks (the rates the mart fast-path otherwise
showed as 0): ``total_cache_read_messages`` (cache.hit_rate),
``total_commands_followed_by_interruption`` (interruption_rate),
``total_command_tools`` / ``total_command_steps`` (avg tools/steps per
command), and ``total_records`` / ``total_errors`` / ``errors_by_category``
(errors total / rate / by_category). These come from the SAME enricher +
``aggregator._command_analysis`` / ``_CacheCollector`` / ``_ErrorsCollector``
logic the full pipeline runs, so they're identical to ``get_project_stats``
(proven by the equivalence tests). Rates are derived at read time from these
counts so a slug's per-provider rows stay additive.
"""

from __future__ import annotations

import json
import sqlite3
from collections import Counter

from stackunderflow.stats import aggregator, classifier, enricher
from stackunderflow.stats.classifier import RawEntry
from stackunderflow.stats.enricher import (
    _has_result_block,
    _text_from,
    _tools_from,
)

from .base import MartBuilder

# Interruption markers the aggregator's ``_command_analysis`` excludes from
# the command tally (``aggregator._is_interrupt_text``). ``str.startswith``
# accepts a tuple, so this drives the same prefix test the pipeline runs.
_INTERRUPT_MARKERS = (classifier.INTERRUPT_PREFIX, classifier.INTERRUPT_API)


class ProjectMartBuilder(MartBuilder):
    """Per-project lifetime aggregates."""

    name = "project"

    def refresh(self, conn: sqlite3.Connection, since_event_id: int) -> int:
        max_id = _max_event_id(conn)
        if max_id <= since_event_id:
            return since_event_id

        affected = [
            int(r[0])
            for r in conn.execute(
                "SELECT DISTINCT project_id FROM usage_events "
                "WHERE id > ? AND id <= ?",
                (since_event_id, max_id),
            ).fetchall()
        ]

        conn.execute(
            """
            INSERT OR REPLACE INTO project_mart (
                project_id, provider, slug, display_name,
                first_ts, last_ts,
                total_messages, total_sessions,
                total_input_tokens, total_output_tokens,
                total_cache_read, total_cache_create,
                total_cost_usd
            )
            SELECT
                e.project_id,
                p.provider,
                p.slug,
                p.display_name,
                MIN(e.ts),
                MAX(e.ts),
                COUNT(*),
                COUNT(DISTINCT e.session_id),
                SUM(e.input_tokens),
                SUM(e.output_tokens),
                SUM(e.cache_read_tokens),
                SUM(e.cache_create_tokens),
                SUM(e.cost_usd)
            FROM usage_events e
            JOIN projects p ON p.id = e.project_id
            WHERE e.project_id IN (
                SELECT DISTINCT project_id
                FROM usage_events
                WHERE id > ? AND id <= ?
            )
            GROUP BY e.project_id, p.provider, p.slug, p.display_name
            """,
            (since_event_id, max_id),
        )

        # Second pass: materialise the message-type + command dims from the
        # raw ``messages`` (the INSERT above reset them to DEFAULT 0).
        _refresh_message_dims(conn, affected)

        return max_id

    def rebuild_from_scratch(self, conn: sqlite3.Connection) -> None:
        conn.execute("DELETE FROM project_mart")
        self.refresh(conn, since_event_id=0)


def _max_event_id(conn: sqlite3.Connection) -> int:
    row = conn.execute("SELECT MAX(id) AS m FROM usage_events").fetchone()
    if row is None:
        return 0
    val = row["m"] if hasattr(row, "keys") else row[0]
    return int(val) if val is not None else 0


def _refresh_message_dims(
    conn: sqlite3.Connection, project_ids: list[int]
) -> None:
    """Recompute the message-type + command + rate dims for *project_ids*.

    Scans every ``messages`` row of each project (joined via ``sessions``,
    ordered by timestamp so the interaction grouping matches the pipeline)
    and classifies it with the same functions ``get_project_stats`` runs,
    then UPDATEs the materialised counts onto the project's mart row. A
    project with no mart row (no billable events) is silently skipped — the
    UPDATE matches nothing.

    Two derivations share the single ``messages`` fetch:

    * ``_count_message_dims`` — the v022 per-message-type + command counts.
    * ``_count_interaction_dims`` — the v023 Overview rate numerators
      (cache-read messages, interruption / tools / steps per command,
      errors total + by_category), which need the full enricher +
      ``aggregator._command_analysis`` pass to match the pipeline exactly.
    """
    for pid in project_ids:
        rows = conn.execute(
            "SELECT m.raw_json AS raw_json, s.session_id AS session_id, "
            "       m.timestamp AS timestamp, p.provider AS provider "
            "FROM messages m "
            "JOIN sessions s ON s.id = m.session_fk "
            "JOIN projects p ON p.id = s.project_id "
            "WHERE s.project_id = ? "
            "ORDER BY m.timestamp",
            (pid,),
        ).fetchall()
        dims = _count_message_dims(r["raw_json"] for r in rows)
        rate = _count_interaction_dims(rows)
        conn.execute(
            "UPDATE project_mart SET "
            "total_user_messages = ?, "
            "total_assistant_messages = ?, "
            "total_tool_use_messages = ?, "
            "total_tool_result_messages = ?, "
            "total_commands = ?, "
            "total_records = ?, "
            "total_errors = ?, "
            "errors_by_category = ?, "
            "total_cache_read_messages = ?, "
            "total_commands_followed_by_interruption = ?, "
            "total_command_tools = ?, "
            "total_command_steps = ? "
            "WHERE project_id = ?",
            (
                dims["user"],
                dims["assistant"],
                dims["tool_use"],
                dims["tool_result"],
                dims["commands"],
                rate["records"],
                rate["errors_total"],
                rate["errors_by_category_json"],
                rate["cache_read_messages"],
                rate["commands_followed_by_interruption"],
                rate["command_tools"],
                rate["command_steps"],
                pid,
            ),
        )


def _count_message_dims(raw_jsons) -> dict[str, int]:
    """Count message-type + command dims over an iterable of ``raw_json`` strings.

    Mirrors ``aggregator.summarise`` exactly by reusing the classifier's
    ``_determine_kind`` and the enricher's field extraction:

    * ``user`` / ``assistant`` == ``overview.message_types`` kind counts.
    * ``tool_use`` == assistant records carrying ``tool_use`` blocks
      (``message_dict['type']=='assistant' and message_dict['tools']``).
    * ``tool_result`` == records carrying a ``tool_result`` block
      (``message_dict['has_tool_result']``).
    * ``commands`` == ``user_interactions.user_commands_analyzed``: a real
      user turn — kind ``user``, not a tool_result, not an interruption.

    Defensive against unparseable rows (a poison row must never break the
    mart refresh / ingest): a row whose ``raw_json`` won't decode to a dict
    is skipped, same coverage the per-message marts already tolerate.
    """
    user = assistant = tool_use = tool_result = commands = 0
    for rj in raw_jsons:
        try:
            payload = json.loads(rj) if rj else {}
        except (json.JSONDecodeError, TypeError, ValueError):
            continue
        if not isinstance(payload, dict):
            continue
        kind = classifier._determine_kind(payload)
        raw_msg = payload.get("message")
        msg = raw_msg if isinstance(raw_msg, dict) else {}
        has_tool_result = _has_result_block(msg)
        if has_tool_result:
            tool_result += 1
        if kind == "user":
            user += 1
            if not has_tool_result and not _text_from(payload).startswith(
                _INTERRUPT_MARKERS
            ):
                commands += 1
        elif kind == "assistant":
            assistant += 1
            if _tools_from(msg):
                tool_use += 1
    return {
        "user": user,
        "assistant": assistant,
        "tool_use": tool_use,
        "tool_result": tool_result,
        "commands": commands,
    }


def _count_interaction_dims(rows) -> dict[str, object]:
    """Materialise the v023 Overview rate numerators for one project.

    Rebuilds the project's ``EnrichedDataset`` the SAME way
    ``queries.build_enriched_dataset`` does — ``classifier.tag`` over
    ``RawEntry`` rows (clean column timestamp wins over any raw payload ts),
    then ``enricher.build`` — and reads the numerators straight off the
    records / interactions and ``aggregator._command_analysis``. Reusing the
    real pipeline functions keeps these counts byte-for-byte identical to
    ``get_project_stats`` (the equivalence tests pin it):

    * ``cache_read_messages`` == ``_CacheCollector.w_read`` — assistant
      records carrying cache-read tokens; ``cache.hit_rate`` numerator.
    * ``errors_total`` / ``errors_by_category_json`` == ``_ErrorsCollector``
      ``_total`` / ``by_category`` — falsy ``error_category`` buckets to
      ``"Other"`` exactly as the collector does.
    * ``commands_followed_by_interruption`` / ``command_tools`` /
      ``command_steps`` == ``_command_analysis``'s
      ``commands_followed_by_interruption`` / ``total_tools_used`` /
      ``total_assistant_steps`` — the interruption_rate and avg
      tools/steps-per-command numerators.
    * ``records`` == ``len(EnrichedDataset.records)`` — the all-kinds record
      count the aggregator's ``errors.rate`` divides by (distinct from
      ``project_mart.total_messages``, which is the billable-event count).

    Defensive against unparseable rows (parity with ``_count_message_dims``):
    a ``raw_json`` that won't decode to a dict is skipped rather than
    breaking the mart refresh.
    """
    raw_entries: list[RawEntry] = []
    for r in rows:
        rj = r["raw_json"]
        try:
            payload = json.loads(rj) if rj else {}
        except (json.JSONDecodeError, TypeError, ValueError):
            continue
        if not isinstance(payload, dict):
            continue
        if r["timestamp"]:
            payload["timestamp"] = r["timestamp"]
        raw_entries.append(
            RawEntry(
                payload=payload,
                session_id=r["session_id"] or "",
                origin=r["session_id"] or "",
                provider=r["provider"] or "anthropic",
            )
        )

    dataset = enricher.build(classifier.tag(raw_entries), "")
    records = dataset.records

    cache_read_messages = sum(
        1
        for rec in records
        if rec.kind == "assistant" and rec.tokens.get("cache_read", 0)
    )

    errors_total = 0
    by_category: Counter[str] = Counter()
    for rec in records:
        if rec.is_error:
            errors_total += 1
            by_category[rec.error_category or "Other"] += 1

    ca = aggregator._command_analysis(records, dataset.interactions)

    return {
        "records": len(records),
        "cache_read_messages": cache_read_messages,
        "errors_total": errors_total,
        "errors_by_category_json": json.dumps(dict(by_category)),
        "commands_followed_by_interruption": int(
            ca.get("commands_followed_by_interruption", 0) or 0
        ),
        "command_tools": int(ca.get("total_tools_used", 0) or 0),
        "command_steps": int(ca.get("total_assistant_steps", 0) or 0),
    }

File: rust/crates/stax-etl/src/marts/project.rs
//! Port of `stackunderflow/etl/marts/project.py` — one row per project.
//!
//! Three passes, in this order, and the order is the design:
//!
//! 1. **Events pass** — recompute the lifetime aggregate of every project with
//!    events in the window, `INSERT OR REPLACE` on the `project_id` primary key.
//!    The statement lists only the original 13 columns, so it *resets the dim
//!    columns to their DEFAULT* on every refresh.
//! 2. **Coverage seed** — a project with zero billable events can never win a
//!    row from a query driven `FROM usage_events`, so it stayed invisible to
//!    every mart-backed read path. Not a rare corner: a Claude `legacy-` history
//!    pseudo-session (user turns only, no token accounting), a provider whose
//!    adapter deliberately emits no usage events, and a normalizer that guards a
//!    row away all produce real projects with real messages and no events.
//! 3. **Dims pass** — [`super::dims::refresh_message_dims`] fills the twelve
//!    columns pass 1 just zeroed, for exactly the affected ids.
//!
//! # The seed's affected-guard is the load-bearing part
//!
//! [`seed_uncovered_projects`] returns **only** the ids it actually inserted,
//! computed by an anti-join *before* the INSERT. An already-covered project must
//! never re-enter `affected`, or every watcher cycle would re-run the full
//! `messages` scan of the dims pass for all 300-odd projects. On steady state
//! the seed contributes nothing and costs one indexed anti-join. The July
//! campaign built it this way deliberately; a port that recomputed `missing`
//! after the INSERT, or fed the INSERT's own row set into `affected`, would be
//! correct and unusably slow.

use anyhow::Result;
use rusqlite::Connection;

use super::{MartBuilder, max_event_id};

/// Per-project lifetime aggregates.
pub struct ProjectMartBuilder;

impl MartBuilder for ProjectMartBuilder {
    fn name(&self) -> &'static str {
        "project"
    }

    fn refresh(&self, conn: &Connection, since_event_id: i64) -> Result<i64> {
        let max_id = max_event_id(conn)?;
        // `affected` drives the second (message-dims) pass. Only ids whose mart
        // row this call actually (re)wrote may enter it.
        let mut affected: Vec<i64> = Vec::new();
        if max_id > since_event_id {
            affected = refresh_from_events(conn, since_event_id, max_id)?;
        }

        // Coverage seed — runs even when no new events arrived, so a project
        // that will never produce an event cannot stay invisible until some
        // unrelated event happens to land.
        affected.extend(seed_uncovered_projects(conn)?);

        super::dims::refresh_message_dims(conn, &affected)?;

        Ok(max_id.max(since_event_id))
    }

    fn rebuild_from_scratch(&self, conn: &Connection) -> Result<()> {
        // DELETE then a from-zero refresh, which runs the coverage seed too —
        // so a rebuild lands the same project set as the incremental path.
        conn.execute("DELETE FROM project_mart", [])?;
        self.refresh(conn, 0)?;
        Ok(())
    }
}

/// `project._refresh_from_events` — returns the affected project ids.
fn refresh_from_events(conn: &Connection, since_event_id: i64, max_id: i64) -> Result<Vec<i64>> {
    let affected: Vec<i64> = {
        let mut stmt =
            conn.prepare("SELECT DISTINCT project_id FROM usage_events WHERE id > ? AND id <= ?")?;
        stmt.query_map(rusqlite::params![since_event_id, max_id], |r| r.get(0))?
            .collect::<Result<Vec<i64>, _>>()?
    };

    conn.execute(
        r"
        INSERT OR REPLACE INTO project_mart (
            project_id, provider, slug, display_name,
            first_ts, last_ts,
            total_messages, total_sessions,
            total_input_tokens, total_output_tokens,
            total_cache_read, total_cache_create,
            total_cost_usd
        )
        SELECT
            e.project_id,
            p.provider,
            p.slug,
            p.display_name,
            MIN(e.ts),
            MAX(e.ts),
            COUNT(*),
            COUNT(DISTINCT e.session_id),
            SUM(e.input_tokens),
            SUM(e.output_tokens),
            SUM(e.cache_read_tokens),
            SUM(e.cache_create_tokens),
            SUM(e.cost_usd)
        FROM usage_events e
        JOIN projects p ON p.id = e.project_id
        WHERE e.project_id IN (
            SELECT DISTINCT project_id
            FROM usage_events
            WHERE id > ? AND id <= ?
        )
        GROUP BY e.project_id, p.provider, p.slug, p.display_name
        ",
        rusqlite::params![since_event_id, max_id],
    )?;
    Ok(affected)
}

/// `project._seed_uncovered_projects` — a zero-cost row for every uncovered
/// project, returning **only** the ids this call inserted.
///
/// The ids come from an anti-join computed *before* the INSERT, so a
/// steady-state call returns `[]` and never re-enters a project into the dims
/// pass. When the anti-join finds nothing the write is skipped entirely.
pub fn seed_uncovered_projects(conn: &Connection) -> Result<Vec<i64>> {
    let missing: Vec<i64> = {
        let mut stmt = conn.prepare(
            "SELECT p.id FROM projects p \
             LEFT JOIN project_mart m ON m.project_id = p.id \
             WHERE m.project_id IS NULL",
        )?;
        stmt.query_map([], |r| r.get(0))?
            .collect::<Result<Vec<i64>, _>>()?
    };
    if missing.is_empty() {
        return Ok(Vec::new());
    }
    // The INSERT deliberately offers every project and lets OR IGNORE drop the
    // covered ones — the anti-join above is what the caller learns from, not
    // this statement's row set.
    conn.execute(
        r"
        INSERT OR IGNORE INTO project_mart (
            project_id, provider, slug, display_name
        )
        SELECT p.id, p.provider, p.slug, p.display_name
        FROM projects p
        ",
        [],
    )?;
    Ok(missing)
}

#[cfg(test)]
mod tests {
    use super::super::testdb;
    use super::*;

    fn user_row(text: &str) -> String {
        format!(r#"{{"type":"human","message":{{"role":"user","content":"{text}"}}}}"#)
    }

    #[test]
    fn a_project_with_events_gets_totals_and_dims() {
        let c = testdb::conn();
        testdb::project(&c, 1, "alpha", "claude");
        testdb::session(&c, 1, 1, "s1");
        testdb::message(
            &c,
            1,
            1,
            0,
            "2026-01-01T00:00:00Z",
            "user",
            "go",
            "[]",
            &user_row("go"),
        );
        testdb::message(
            &c,
            2,
            1,
            1,
            "2026-01-01T00:01:00Z",
            "assistant",
            "ok",
            "[]",
            r#"{"type":"assistant","message":{"role":"assistant","content":[
                {"type":"tool_use","id":"a","name":"Read","input":{}}]}}"#,
        );
        testdb::event(
            &c,
            1,
            Some(2),
            1,
            "s1",
            "claude",
            "m",
            "2026-01-01",
            (10, 5, 2, 1),
            1.25,
        );

        ProjectMartBuilder.refresh(&c, 0).unwrap();

        let (msgs, sess, cost, u, a, tu, cmds, steps, tools): (
            i64,
            i64,
            f64,
            i64,
            i64,
            i64,
            i64,
            i64,
            i64,
        ) = c
            .query_row(
                "SELECT total_messages, total_sessions, total_cost_usd, total_user_messages, \
                 total_assistant_messages, total_tool_use_messages, total_commands, \
                 total_command_steps, total_command_tools FROM project_mart",
                [],
                |r| {
                    Ok((
                        r.get(0)?,
                        r.get(1)?,
                        r.get(2)?,
                        r.get(3)?,
                        r.get(4)?,
                        r.get(5)?,
                        r.get(6)?,
                        r.get(7)?,
                        r.get(8)?,
                    ))
                },
            )
            .unwrap();
        assert_eq!((msgs, sess), (1, 1));
        assert!((cost - 1.25).abs() < 1e-12);
        assert_eq!((u, a, tu, cmds), (1, 1, 1, 1));
        assert_eq!((steps, tools), (1, 1));
    }

    #[test]
    fn an_event_less_project_is_seeded_with_zero_totals_and_real_dims() {
        // The coverage case: a history-only project. Totals are truthfully
        // zero; the dims still report its user-message and command counts.
        let c = testdb::conn();
        testdb::project(&c, 7, "history-only", "claude");
        testdb::session(&c, 1, 7, "legacy-s1");
        testdb::message(
            &c,
            1,
            1,
            0,
            "2026-01-01T00:00:00Z",
            "user",
            "a",
            "[]",
            &user_row("a"),
        );
        testdb::message(
            &c,
            2,
            1,
            1,
            "2026-01-01T00:01:00Z",
            "user",
            "b",
            "[]",
            &user_row("b"),
        );

        ProjectMartBuilder.refresh(&c, 0).unwrap();

        let (pid, slug, msgs, cost, u, cmds): (i64, String, i64, f64, i64, i64) = c
            .query_row(
                "SELECT project_id, slug, total_messages, total_cost_usd, \
                 total_user_messages, total_commands FROM project_mart",
                [],
                |r| {
                    Ok((
                        r.get(0)?,
                        r.get(1)?,
                        r.get(2)?,
                        r.get(3)?,
                        r.get(4)?,
                        r.get(5)?,
                    ))
                },
            )
            .unwrap();
        assert_eq!((pid, slug.as_str()), (7, "history-only"));
        assert_eq!(msgs, 0);
        assert!(cost.abs() < f64::EPSILON);
        assert_eq!((u, cmds), (2, 2));
    }

    #[test]
    fn the_seed_returns_only_ids_it_inserted() {
        // The anti-join guard. If the seed reported already-covered ids, every
        // steady-state refresh would re-scan every project's messages.
        let c = testdb::conn();
        testdb::project(&c, 1, "a", "claude");
        testdb::project(&c, 2, "b", "claude");
        let first = seed_uncovered_projects(&c).unwrap();
        assert_eq!(first, vec![1, 2]);
        let second = seed_uncovered_projects(&c).unwrap();
        assert!(second.is_empty(), "steady state must contribute nothing");

        // …and a newly appearing project is the only one reported.
        testdb::project(&c, 3, "c", "claude");
        assert_eq!(seed_uncovered_projects(&c).unwrap(), vec![3]);
    }

    #[test]
    fn the_events_pass_resets_the_dim_columns_and_the_second_pass_refills_them() {
        let c = testdb::conn();
        testdb::project(&c, 1, "a", "claude");
        testdb::session(&c, 1, 1, "s1");
        testdb::message(
            &c,
            1,
            1,
            0,
            "2026-01-01T00:00:00Z",
            "user",
            "go",
            "[]",
            &user_row("go"),
        );
        testdb::message(
            &c,
            2,
            1,
            1,
            "2026-01-01T00:01:00Z",
            "assistant",
            "ok",
            "[]",
            r#"{"type":"assistant","message":{"role":"assistant","content":"ok"}}"#,
        );
        testdb::event(
            &c,
            1,
            Some(2),
            1,
            "s1",
            "claude",
            "m",
            "2026-01-01",
            (1, 1, 0, 0),
            0.5,
        );
        ProjectMartBuilder.refresh(&c, 0).unwrap();

        // A second window over the same project re-writes the row (zeroing the
        // dims) and must refill them.
        testdb::event(
            &c,
            2,
            None,
            1,
            "s1",
            "claude",
            "m",
            "2026-01-02",
            (1, 1, 0, 0),
            0.5,
        );
        ProjectMartBuilder.refresh(&c, 1).unwrap();

        let (msgs, u, cmds): (i64, i64, i64) = c
            .query_row(
                "SELECT total_messages, total_user_messages, total_commands FROM project_mart",
                [],
                |r| Ok((r.get(0)?, r.get(1)?, r.get(2)?)),
            )
            .unwrap();
        assert_eq!(msgs, 2, "the totals are recomputed from ALL events");
        assert_eq!((u, cmds), (1, 1), "the dims must not be left at DEFAULT 0");
    }

    #[test]
    fn refresh_returns_the_high_water_mark_even_with_nothing_new() {
        let c = testdb::conn();
        testdb::project(&c, 1, "a", "claude");
        testdb::event(
            &c,
            5,
            None,
            1,
            "s1",
            "claude",
            "m",
            "2026-01-01",
            (1, 1, 0, 0),
            0.0,
        );
        assert_eq!(ProjectMartBuilder.refresh(&c, 0).unwrap(), 5);
        // `max(max_id, since_event_id)` — project is the one mart that does not
        // early-return, because the coverage seed must still run.
        assert_eq!(ProjectMartBuilder.refresh(&c, 9).unwrap(), 9);
    }

    #[test]
    fn a_rebuild_lands_the_same_project_set_as_the_incremental_path() {
        let c = testdb::conn();
        testdb::project(&c, 1, "with-events", "claude");
        testdb::project(&c, 2, "without", "claude");
        testdb::event(
            &c,
            1,
            None,
            1,
            "s1",
            "claude",
            "m",
            "2026-01-01",
            (1, 1, 0, 0),
            1.0,
        );
        ProjectMartBuilder.refresh(&c, 0).unwrap();
        let before: i64 = c
            .query_row("SELECT COUNT(*) FROM project_mart", [], |r| r.get(0))
            .unwrap();
        ProjectMartBuilder.rebuild_from_scratch(&c).unwrap();
        let after: i64 = c
            .query_row("SELECT COUNT(*) FROM project_mart", [], |r| r.get(0))
            .unwrap();
        assert_eq!((before, after), (2, 2));
    }
}
